unitary/alpha/sparse_vector_simulator.py

```
@@ -0,0 +1,121 @@
+            mask = ~0
+            for dst_bit, qubit in enumerate(qubits[::-1]):
+                src_bit = self.qubit_map[qubit]
+                bit = np.array((self._states >> src_bit) & 1, dtype=int)
+                dst_rows |= bit << dst_bit
+                reverse_affected = np.concatenate(
+                    (reverse_affected, reverse_affected | 1 << src_bit)
+                )
+                mask ^= 1 << src_bit
+            self._states &= mask
+            unique_unaffected, dst_cols = np.unique(self._states, return_inverse=True)
+            partitioned_state = np.zeros(
+                (1 << len(qubits), len(unique_unaffected)), np.complex128
+            )
+            partitioned_state[dst_rows, dst_cols] = self._amplitudes
+            np.matmul(cirq.unitary(action), partitioned_state, out=partitioned_state)
+            nz_rows, nz_cols = np.nonzero(abs(partitioned_state) > EPSILON)
+            self._states = reverse_affected[nz_rows] | unique_unaffected[nz_cols]
+            self._amplitudes = partitioned_state[nz_rows, nz_cols]
+        return True
+
+    def _perform_measurement(self, qubits):
+        raise NotImplementedError
+
+    # abstract method of OperationTarget
+    def sample(self, qubits, repetitions, prng):
+        probs = abs(self._amplitudes) ** 2
+        samples = prng.choice(self._states, size=repetitions, p=probs)
+        out = np.empty((repetitions, len(qubits)), dtype=np.uint8)
+        for j, q in enumerate(qubits):
+            out[:, j] = (samples >> self.qubit_map[q]) & 1
+        return out
+
+    def post_select(self, qubit, value):
+        assert value in (0, 1)
+        mask = 1 << self.qubit_map[qubit]
+        value *= mask
+        nonzero_indices = np.nonzero(self._states & mask == value)
+        self._states = self._states[nonzero_indices]
+        self._amplitudes = self._amplitudes[nonzero_indices]
+        self._amplitudes /= np.linalg.norm(self._amplitudes)
```

What happens if a value is post_selected that is invalid (ie. the norm is zero)?  Maybe we should raise an error here?

---

Added an exception InvalidPostSelectionError.